File: entities/xotaker.js
class Xotaker extends Attacker {
    constructor() {
        super();
    }

    _nextTick() {
        if (this.isNear(Air) && this.canMultiply()) {
            let cell = random(this.findNear(Air));
            this.multiply(cell);
        }

        else if (this.isNear(Grass) && this.canMultiply() && this.energy == this._energyLimit) {
            let cell = random(this.findNear(Grass));
            let oldPos = this.pos;
            this.eat(cell);
            this.multiply(oldPos);
        }

        else if (this.isNear(Grass)) {
            let cell = random(this.findNear(Grass))
            this.eat(cell);
        }

        else if (this.isNear(Air)) {
            let cell = random(this.findNear(Air));
            this.move(cell)
        }
    }
}

Xotaker._defaultEnergy = 2;
Xotaker._multiplicationCost = 2;
Xotaker._energyLimit = 4;
Xotaker._movementCost = 0.1;
Xotaker._energyFromPrey = 0.2;

File: entities/xotaker.py
from random import choice

from entities.air import Air
from entities.grass import Grass
from entities.attacker import Attacker


class Xotaker(Attacker):
    __slots__ = ()

    _default_energy = 2
    energy_limit = 4
    _energy_increment_per_tick = 0
    _movement_cost = 0.1
    _energy_from_prey = 0.2

    def _next_tick(self):
        if self._do_multiply_on(Air) and self._is_near(Air):
            pos = choice(self._find_near(Air)).pos
            self._multiply(pos)

        elif self._do_multiply_on(Grass) and self._is_near(Grass):
            pos = choice(self._find_near(Grass)).pos
            old_pos = self.pos
            self._eat(pos)
            self._multiply(old_pos)

        elif self._is_near(Grass):
            pos = choice(self._find_near(Grass)).pos
            self._eat(pos)

        elif self._is_near(Air):
            pos = choice(self._find_near(Air)).pos
            self._move(pos)

    def _do_multiply_on(self, kind):
        if kind is Grass:
            return self.energy >= 2*self._default_energy
        return super()._do_multiply_on(kind)
